utils/near-performance-metrics/src/stats.rs

```diff
@@ -0,0 +1,341 @@
+
+    fn log(&mut self, class_name: &'static str, msg: &'static str, line: u32, took: Duration) {
+        let entry = self.stat.entry((msg, line)).or_insert_with(|| Entry {
+            cnt: 0,
+            time: Duration::default(),
+            max_time: Duration::default(),
+        });
+        entry.cnt += 1;
+        entry.time += took;
+        entry.max_time = max(took, entry.max_time);
+
+        self.cnt += 1;
+        self.time += took;
+        self.classes.insert(class_name);
+    }
+
+    fn print_stats(&self, tid: usize, sleep_time: Duration) {
+        let ratio = (self.time.as_nanos() as f64) / (sleep_time.as_nanos() as f64);
+
+        let class_name = format!("{:?}", self.classes);
+
+        if ratio > MIN_OCCUPANCY_RATIO_THRESHOLD {
+            info!(
+                "    Thread:{} {}:{} ratio: {:.4}",
+                tid,
+                class_name,
+                TID.with(|x| *x.borrow()),
+                ratio
+            );
+            for entry in self.stat.iter() {
+                info!(
+                    "        func {} {}:{} cnt: {} total: {}ms max: {}ms",
+                    TID.with(|x| *x.borrow()),
+                    (entry.0).0,
+                    (entry.0).1,
+                    entry.1.cnt,
+                    ((entry.1.time.as_millis()) as f64),
+                    ((entry.1.max_time.as_millis()) as f64)
+                );
+            }
+        }
+    }
+}
+
+struct Stats {
+    stats: HashMap<usize, ThreadStats>,
+}
+
+impl Stats {
+    fn new() -> Self {
+        Self { stats: HashMap::new() }
+    }
+
+    fn log(&mut self, class_name: &'static str, msg: &'static str, line: u32, took: Duration) {
+        let tid = TID.with(|x| {
+            if *x.borrow_mut() == 0 {
+                *x.borrow_mut() = NTHREADS.fetch_add(1, Ordering::SeqCst);
+                self.stats.insert(*x.borrow_mut(), ThreadStats::new());
+            }
+            *x.borrow_mut()
+        });
+        if let Some(val) = self.stats.get_mut(&tid) {
+            val.log(class_name, msg, line, took)
+        }
+    }
+
+    fn print_stats(&self, sleep_time: Duration) {
+        info!("Performance stats {} threads", self.stats.len());
+        for entry in self.stats.iter() {
+            entry.1.print_stats(*entry.0, sleep_time);
+        }
+    }
+}
+
+pub fn measure_performance<F, Message, Result>(
+    class_name: &'static str,
+    msg: Message,
+    f: F,
+) -> Result
+where
+    F: FnOnce(Message) -> Result,
+{
+    let now = Instant::now();
+    let result = f(msg);
+
+    let took = now.elapsed();
+    if took > SLOW_CALL_THRESHOLD {
+        info!(
+            "SLOW {}:{} {:?} took: {}ms",
+            class_name,
+            TID.with(|x| *x.borrow()),
+            std::any::type_name::<Message>(),
+            took.as_millis(),
+        );
+    }
+
+    STATS.write().unwrap().log(class_name, std::any::type_name::<Message>(), 0, took);
+    result
+}
+
+pub fn measure_performance2<F, Result>(
```

Comment:
Where is it used? Let's have a better name for it :smile:

Reply:
I removed that method, it's not longer needed.